Rehaul: List widget (#896)
appsmithorg/appsmith-docs: website/docs/reference/widgets/list.md

# List

List widget provides a way to iterate over a structured dataset(array of objects) and display the data in repeating sections without writing any code. Each list item can contain other widgets to display data or capture user input.

<VideoEmbed host="youtube" videoId="0ePiZlWmp7Q" title="How to use List Widget" caption="How to use List Widget"/>

:::info

> Currenlty,you can only use widgets to display the data like images, text in the list.

Reviewer: ```suggestion
Although you can add many other widgets inside the List widget, currently, Appsmith fully supports only the Text and Image widgets inside a List.
```

:::

##  List components

> Typically, a list widget is a collection of other widgets that you can embed to display data or capture user input. Once you drag the list widget on the canvas, you can see it has embedded widgets [ image](./image.md) and two[ texts](./text.md) available. You can add more widgets as per your requirement. Add the widgets only on the first list item container. The widgets get replicated for the other list items automatically when a dataset is connected to the list. 

Reviewer: ```suggestion
Typically, a List widget is a collection of other widgets you can embed to display data or capture user input. Once you drag a List widget on the canvas, it has embedded Image and Text widgets by default. You can add more widgets per your requirement on the first list item container. The widgets get replicated on the other list items automatically when a dataset is connected to the list. 
```

#### Dynamic data mapping

If you wish to bind the dynamic response of the queries, then you can use mustache syntax (`{{ }}`) to embed the same. For example, You have written a query `“GetAllEmployees”` then to bind the response use `{{GetAllEmployees.data}}`. Typically, you have a format of `{{QUERY_NAME.data}}` to bind the data returned by the query.
If you are binding the dynamic response of your query or API to the list widget, remember to use the query columns or API response object’s attributes to map to the individual widget by using `currentItem.<attribute_or_column_name>`.

### Server-side pagination

Lists are often required to display large data sets from queries, but browsers can't always load all the data present in the database, or might do so slowly. You can use server-side pagination when a client receives only a subset of data from large datasets. It allows you to define the data limit that a query call can render. Thus, allowing you to paginate the data and determine the pagination boundaries.

Follow the steps below to paginate the responses and request smaller chunks of data at a time:

> 1. Enable the server-side pagination property in the list

Reviewer: ```suggestion
1. Enable the Server Side Pagination property for the List
```

> 2. Call the query via the onPageChange event.

Reviewer: ```suggestion
2. Call the query on the onPageChange event listener.
```

> 3. Calculate the required pageOffset using the pageSize and pageNo property as shown below:

Reviewer: ```suggestion
```

> ```javascript

Reviewer: Remove this code block

List1.pageOffset = (List1.pageNo - 1) * List1.pageSize
```

> 4. Set the LIMIT and OFFSET in the query:

Reviewer: ```suggestion
4. Set the LIMIT and OFFSET clause in the query:
```

```javascript

> SELECT * FROM users LIMIT {{ List1.pageSize }} OFFSET {{ List1.pageOffset }}

Reviewer: ```suggestion
SELECT * FROM users LIMIT {{ List1.pageSize }} OFFSET {{ (List1.pageNo - 1) * List1.pageSize }}
```

```

## Properties

> Properties allow you to edit the widget, connect it with other widgets and customize the user actions.

Reviewer: ```suggestion
Properties allow you to customize the widget, connect it to other widgets and trigger events on user actions.
```

### Widget properties

> These properties allow you to name, edit the List widget. All these properties are present in the property pane of the widget. The following table lists all the widget properties.

Reviewer: remove this line

| **Items**                  | Allows you to bind static or dynamic data collection to the widget.                                                                                                                                                                                       |
| **Server-side Pagination** | Enables you to implement pagination by limiting the number of results fetched per Query request.  |
| **Visible**                | Controls widget's visibility on the page. When turned off: The widget is visible when the app is published. It appears translucent when in Edit mode.                                                                                            |
| **Animate Loading**        | When turned off, the widget loads without any skeletal animation. You can use a toggle switch to turn it on/off. You can also turn it off/on using javascript by enabling the JS label next to it.                                                    |

### Reference properties

> These properties allow you to bind your List widget with any other widget in queries or JS objects. The following table lists all the binding properties.

Reviewer: ```suggestion
These properties can be referenced in other widgets , queries, or JS functions using the dot operator.
```

### Binding Properties
### Reference properties

These properties allow you to bind your List widget with any other widget in queries or JS objects. The following table lists all the binding properties.

| Binding Property    | Description                                                                                                                                                                                                                               |
| Reference Property    | Description                                                                                                                                                                                                                               |
| ------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **backgroundColor** | Represents the widget's **Background Color** setting as a CSS `color` value _(string)_.                                                                                                                                                   |

Reviewer: Both these properties are reference properties as shown in screenshot below. keep them here
<img width="205" alt="Screenshot 2023-02-22 at 14 05 46" src="https://user-images.githubusercontent.com/32227135/220566351-cc02415c-bd9a-46e7-a426-b5a918c6bd5b.png">

> ### Styles

Reviewer: ```suggestion
### Style properties
```

> | ------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

Reviewer: Change the heading of the column to just 'Property' instead of  'Style property'
